src/Controller/User/Profile/User2FAController.php

```diff
@@ -174,27 +176,38 @@ private function handleForm(
     ): FormInterface|Response {
         $form->handleRequest($request);
 
-        if ($form->isSubmitted()
-                && $form->has('totpCode')
+        if (!$form->isSubmitted()) {
+            return $form;
+        }
+
+        if ($form->has('totpCode')
                 && !$this->setupHasValidCode($dto->totpSecret, $form->get('totpCode')->getData())) {
             $form->get('totpCode')->addError(new FormError($this->translator->trans('2fa.code_invalid')));
-        }
 
-        if ($form->isSubmitted() && $form->isValid()) {
```

Comment:
this was the key, as `$form->isSubmitted() && $form->isValid()` the form was submitted, but not valid, in which case it just dropped all the way to the end returning the form with no other logic handling